### PR TITLE
Replace outdated httpClient with OkHttp

### PR DESCRIPTION
**Pull request by vernazza**
Library httpClient is deprecated and outdated, but it worked since MS changed some certificates server side.
The new certificates are not supported by the cognitive-face-android library and the library can't work anymore.
This pull request replaces httpClient with OkHttp library

### CODE REVIEW

**Comment by lebronJ** [comment]
Hi @vernazza , thank you so much for your contribution to Face Android SDK! Actually we have already spent some time to test with your implementation and review the code, and it works like a charm! 

But since we are working on some internal urgent tasks, we need some more time before merging this. Will work on this by end of this week, thanks!

File: ClientLibrary/lib/build.gradle [hunk 1]
```diff
@@ -2,19 +2,25 @@ apply plugin: 'com.android.library'
...
+    }
 }
 
 dependencies {
-    implementation 'com.google.code.gson:gson:2.8.2'
+    implementation 'com.google.code.gson:gson:2.8.5'
+    implementation 'com.squareup.okhttp3:okhttp:3.13.1'
 }
 
+
```
Comment (lebronJ) [inline ClientLibrary/lib/build.gradle]: [Nit] Remove this.

File: ClientLibrary/lib/build.gradle [hunk 2]
```diff
@@ -29,6 +35,7 @@ archivesBaseName = "face"
 version = "1.4.3"
 
 // Upload artifacts to maven central repository staging servers
+
```
Comment (lebronJ) [inline ClientLibrary/lib/build.gradle]: [Nit] Remove this.

File: ClientLibrary/lib/build.gradle [hunk 3]
```diff
@@ -80,3 +87,4 @@ uploadArchives {
         }
     }
 }
+*/
```
Comment (lebronJ) [inline ClientLibrary/lib/build.gradle]: [Nit] Remove this.

File: ClientLibrary/lib/src/main/java/com/microsoft/projectoxford/face/rest/WebServiceRequest.java [hunk 4]
```diff
@@ -96,150 +87,156 @@ public Object request(String url, RequestMethod method, Map<String, Object> data
...
-            throw new ClientException("Error executing GET request!", statusCode);
+            throw new ClientException("Error executing GET request!", response.code());
         }
     }
 
 
     private Object patch(String url, Map<String, Object> data, String contentType) throws ClientException, IOException {
-        HttpPatch request = new HttpPatch(url);
-        request.setHeader(HEADER_KEY, mSubscriptionKey);
+
```
Comment (lebronJ) [inline ClientLibrary/lib/src/main/java/com/microsoft/projectoxford/face/rest/WebServiceRequest.java]: [Nit] Remove this.

File: ClientLibrary/lib/src/main/java/com/microsoft/projectoxford/face/rest/WebServiceRequest.java [hunk 5]
```diff
@@ -96,150 +87,156 @@ public Object request(String url, RequestMethod method, Map<String, Object> data
...
-        request.setHeader(CONTENT_TYPE, APPLICATION_JSON);
-        HttpResponse response = mClient.execute(request);
-
-        int statusCode = response.getStatusLine().getStatusCode();
-        if (statusCode == HttpStatus.SC_OK) {
-            return readInput(response.getEntity().getContent());
+
+        Request.Builder builder = new Request.Builder()
+                .url(url)
+                .header(HEADER_KEY, mSubscriptionKey)
```
Comment (lebronJ) [inline ClientLibrary/lib/src/main/java/com/microsoft/projectoxford/face/rest/WebServiceRequest.java]: [Nit] Will `.addHeader(HEADER_KEY, this.mSubscriptionKey)` work? Let's keep it consistent with below `Request.Builder`s.

File: ClientLibrary/lib/src/main/java/com/microsoft/projectoxford/face/rest/WebServiceRequest.java [hunk 6]
```diff
@@ -96,150 +87,156 @@ public Object request(String url, RequestMethod method, Map<String, Object> data
...
+                .patch(RequestBody.create(MediaType.get(APPLICATION_JSON), json))
+                .header(CONTENT_TYPE, APPLICATION_JSON);
```
Comment (lebronJ) [inline ClientLibrary/lib/src/main/java/com/microsoft/projectoxford/face/rest/WebServiceRequest.java]: [Nit] Can we unify the usage of `header` and `addHeader` if they are of same usage? Thanks.

File: ClientLibrary/lib/src/main/java/com/microsoft/projectoxford/face/rest/WebServiceRequest.java (hunk 5 above)
Comment (lebronJ) [reply ClientLibrary/lib/src/main/java/com/microsoft/projectoxford/face/rest/WebServiceRequest.java]: [Nit] Also let's unify the code of `Request.Builder builder` and `Request request = //... .build()`

File: ClientLibrary/lib/src/main/java/com/microsoft/projectoxford/face/rest/WebServiceRequest.java [hunk 7]
```diff
@@ -96,150 +87,156 @@ public Object request(String url, RequestMethod method, Map<String, Object> data
...
-            request.setEntity(new ByteArrayEntity((byte[]) data.get(DATA)));
+            builder.post(RequestBody.create(MediaType.get(contentType), (byte[]) data.get(DATA)));
         }
 
-        HttpResponse response = mClient.execute(request);
-        int statusCode = response.getStatusLine().getStatusCode();
-        if (statusCode == HttpStatus.SC_OK || statusCode == HttpStatus.SC_ACCEPTED) {
-            return readInput(response.getEntity().getContent());
+        Response response = mClient.newCall(builder.build()).execute();
+
```
Comment (lebronJ) [inline ClientLibrary/lib/src/main/java/com/microsoft/projectoxford/face/rest/WebServiceRequest.java]: [Nit] Remove this.

File: ClientLibrary/lib/src/main/java/com/microsoft/projectoxford/face/rest/WebServiceRequest.java [hunk 8]
```diff
@@ -96,150 +87,156 @@ public Object request(String url, RequestMethod method, Map<String, Object> data
...
                 }
             }
 
-            throw new ClientException("Error executing POST request!", statusCode);
+            throw new ClientException("Error executing POST request!", response.code());
         }
     }
 
     private Object put(String url, Map<String, Object> data) throws ClientException, IOException {
-        HttpPut request = new HttpPut(url);
```
Comment (lebronJ) [inline ClientLibrary/lib/src/main/java/com/microsoft/projectoxford/face/rest/WebServiceRequest.java]: [Nit] Remove this blank line.

File: ClientLibrary/lib/src/main/java/com/microsoft/projectoxford/face/rest/WebServiceRequest.java [hunk 9]
```diff
@@ -96,150 +87,156 @@ public Object request(String url, RequestMethod method, Map<String, Object> data
...
+
+        Request.Builder builder = new Request.Builder()
+                .url(url)
+                .header(HEADER_KEY, mSubscriptionKey)
+                .header(CONTENT_TYPE, APPLICATION_JSON)
+                .put(RequestBody.create(MediaType.get(APPLICATION_JSON), json));
+
+        Response response = mClient.newCall(builder.build()).execute();
+
+         if (response.isSuccessful()) {
```
Comment (lebronJ) [inline ClientLibrary/lib/src/main/java/com/microsoft/projectoxford/face/rest/WebServiceRequest.java]: [Nit] Please take care of the indent here.

File: ClientLibrary/lib/src/main/java/com/microsoft/projectoxford/face/rest/WebServiceRequest.java [hunk 10]
```diff
@@ -267,14 +264,10 @@ public static String getUrl(String path, Map<String, Object> params) {
...
-    private String readInput(InputStream is) throws IOException {
-        BufferedReader br = new BufferedReader(new InputStreamReader(is));
-        StringBuffer json = new StringBuffer();
-        String line;
-        while ((line = br.readLine()) != null) {
-            json.append(line);
-        }
+    private String readInput(Response response) throws IOException {
+        if (response.body() == null)
+            return null;
```
Comment (lebronJ) [inline ClientLibrary/lib/src/main/java/com/microsoft/projectoxford/face/rest/WebServiceRequest.java]: [Nit] Let's wrap this with brackets.

File: Sample/app/build.gradle [hunk 11]
```diff
@@ -26,7 +29,7 @@ dependencies {
     // implementation project(':lib')
 
     // Use the following line to include client library for Face API from Maven Central Repository
-    implementation 'com.microsoft.projectoxford:face:1.4.3'
+    //implementation 'com.microsoft.projectoxford:face:1.4.3'
```
Comment (lebronJ) [inline Sample/app/build.gradle]: We will first release a new version of Android SDK and then let's update this to use the public one.

File: ClientLibrary/lib/build.gradle (hunk 1 above)
Comment (vernazza) [reply ClientLibrary/lib/build.gradle]: Yes, that shouldn't have reached the pull request

File: ClientLibrary/lib/build.gradle (hunk 2 above)
Comment (vernazza) [reply ClientLibrary/lib/build.gradle]: Yes, that shouldn't have reached the pull request

File: ClientLibrary/lib/build.gradle (hunk 3 above)
Comment (vernazza) [reply ClientLibrary/lib/build.gradle]: Yes, that shouldn't have reached the pull request

File: ClientLibrary/lib/src/main/java/com/microsoft/projectoxford/face/rest/WebServiceRequest.java (hunk 5 above)
Comment (vernazza) [reply ClientLibrary/lib/src/main/java/com/microsoft/projectoxford/face/rest/WebServiceRequest.java]: addHeader allows having multiple headers with the same key, while header replace existing headers with the same key.
I think that in this situation "header" is better; I'll replace addHeader with header

File: ClientLibrary/lib/src/main/java/com/microsoft/projectoxford/face/rest/WebServiceRequest.java (hunk 6 above)
Comment (vernazza) [reply ClientLibrary/lib/src/main/java/com/microsoft/projectoxford/face/rest/WebServiceRequest.java]: sure

File: ClientLibrary/lib/src/main/java/com/microsoft/projectoxford/face/rest/WebServiceRequest.java (hunk 7 above)
Comment (vernazza) [reply ClientLibrary/lib/src/main/java/com/microsoft/projectoxford/face/rest/WebServiceRequest.java]: What?

File: ClientLibrary/lib/src/main/java/com/microsoft/projectoxford/face/rest/WebServiceRequest.java (hunk 8 above)
Comment (vernazza) [reply ClientLibrary/lib/src/main/java/com/microsoft/projectoxford/face/rest/WebServiceRequest.java]: ok

File: ClientLibrary/lib/src/main/java/com/microsoft/projectoxford/face/rest/WebServiceRequest.java (hunk 4 above)
Comment (vernazza) [reply ClientLibrary/lib/src/main/java/com/microsoft/projectoxford/face/rest/WebServiceRequest.java]: ok

File: Sample/app/build.gradle (hunk 11 above)
Comment (vernazza) [reply Sample/app/build.gradle]: Not sure about this... the sample should then use the local module, isn't it?

File: ClientLibrary/lib/src/main/java/com/microsoft/projectoxford/face/rest/WebServiceRequest.java [hunk 12]
```diff
@@ -96,155 +87,158 @@ public Object request(String url, RequestMethod method, Map<String, Object> data
...
-        StringEntity entity = new StringEntity(json, "utf-8");
-        request.setEntity(entity);
-        request.setHeader(CONTENT_TYPE, APPLICATION_JSON);
-        HttpResponse response = mClient.execute(request);
-
-        int statusCode = response.getStatusLine().getStatusCode();
-        if (statusCode == HttpStatus.SC_OK) {
-            return readInput(response.getEntity().getContent());
+
+        Request.Builder builder = new Request.Builder()
```
Comment (lebronJ) [inline ClientLibrary/lib/src/main/java/com/microsoft/projectoxford/face/rest/WebServiceRequest.java]: [Nit] let's unify by using `Request request = //... .build()` here.

File: Sample/app/build.gradle (hunk 11 above)
Comment (lebronJ) [reply Sample/app/build.gradle]: It should use the publicly released Android SDK on maven. We will first upgrade the SDK to 1.4.4 when the code review is complete, and then let you know, then pls help change this to refer to the 1.4.4. Does it make sense to you?

File: ClientLibrary/lib/src/main/java/com/microsoft/projectoxford/face/rest/WebServiceRequest.java (hunk 5 above)
Comment (lebronJ) [reply ClientLibrary/lib/src/main/java/com/microsoft/projectoxford/face/rest/WebServiceRequest.java]: Agreed.

File: Sample/app/build.gradle (hunk 11 above)
Comment (vernazza) [reply Sample/app/build.gradle]: ok

**Comment by vernazza** [comment]
I think that all the comments have been resolved and applied

**Comment by lebronJ** [comment]
> I think that all the comments have been resolved and applied

Yes, we will have final testing and then release the Android package publicly. Will let you know then. Thanks for your patience and contribution!

**Comment by lebronJ** [comment]
@vernazza The upgraded 1.4.4 version of `Cognitive-Face-Android SDK` has been published. Thanks again for your contribution, we will take care of the sample in next PR. Closed this.

**Pull request merged**